### PR TITLE
[GLUTEN-5096][CH]Bug fix regex extract diff

### PR DESCRIPTION
**Pull request by KevinyhZou**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#5096)

## How was this patch tested?
TEST BY UT

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Review by liuneng1994 — APPROVED** [review]
LGTM

File: cpp-ch/local-engine/Parser/scalar_function_parser/regexp_extract.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            {
+                std::string expr_str = literal_expr.string();
+                size_t expr_size = expr_str.size();
+                if (expr_str.data()[expr_size - 1] == '$')
+                    expr_str.replace(expr_str.find_last_of("$"), 1, "(?:(\n)*)$");
+                
+                auto * regex_expr_node = addColumnToActionsDAG(actions_dag, std::make_shared<DataTypeString>(), expr_str);
+                auto parsed_args = parseFunctionArguments(substrait_func, "", actions_dag);
+                parsed_args[1] =  regex_expr_node;
+                return toFunctionNode(actions_dag, "regexpExtract", parsed_args);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/regexp_extract.cpp:63]: 需要根据output_type转下类型。

File: cpp-ch/local-engine/Parser/scalar_function_parser/regexp_extract.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    static constexpr auto name = "regexp_extract";
+    String getName() const override { return name; }
+
+    const ActionsDAG::Node * parse(
+        const substrait::Expression_ScalarFunction & substrait_func,
+        ActionsDAGPtr & actions_dag) const override
+    {
+        const auto & args = substrait_func.arguments();
+        if (args.size() != 2 && args.size() != 3)
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/regexp_extract.cpp:47]: substrait scalar function中固定三个arg, 不会有两个。

File: cpp-ch/local-engine/Parser/scalar_function_parser/regexp_extract.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} requires 2 or 3 arguments", getName());
+        
+        if(args[1].value().has_literal())
+        {
+            auto literal_expr = args[1].value().literal();
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/regexp_extract.cpp:52]: 传值会导致复制构造开销，const auto & literal_expr = xxx; 能避免。

File: cpp-ch/local-engine/Parser/scalar_function_parser/regexp_extract.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            if (literal_expr.has_string())
+            {
+                std::string expr_str = literal_expr.string();
+                size_t expr_size = expr_str.size();
+                if (expr_str.data()[expr_size - 1] == '$')
+                    expr_str.replace(expr_str.find_last_of("$"), 1, "(?:(\n)*)$");
+                
+                auto * regex_expr_node = addColumnToActionsDAG(actions_dag, std::make_shared<DataTypeString>(), expr_str);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/regexp_extract.cpp:60]: const auto *

**Pull request merged**